src/core/sock/sockinfo_tcp.cpp

```diff
@@ -6042,3 +6029,192 @@ inline bool sockinfo_tcp::handle_bind_no_port(int &bind_ret, in_port_t in_port,
 
     return CONTINUE_WITH_BIND;
 }
+
+int sockinfo_tcp::tcp_tx_express(const struct iovec *iov, unsigned iov_len, uint32_t mkey,
+                                 xlio_express_flags flags, void *opaque_op)
+{
+    if (unlikely(!is_rts())) {
+        if (m_conn_state == TCP_CONN_TIMEOUT) {
+            si_tcp_logdbg("TX timed out");
+            errno = ETIMEDOUT;
+        } else if (m_conn_state == TCP_CONN_RESETED) {
+            si_tcp_logdbg("TX on reseted socket");
+            errno = ECONNRESET;
+        } else if (m_conn_state == TCP_CONN_ERROR) {
+            si_tcp_logdbg("TX on connection failed socket");
+            errno = ECONNREFUSED;
+        } else {
+            si_tcp_logdbg("TX on disconnected socket");
+            errno = EPIPE;
+        }
+        return -1;
+    }
+
+    err_t err;
+    pbuf_desc mdesc;
+
+    switch (flags & XLIO_EXPRESS_OP_TYPE_MASK) {
+    case XLIO_EXPRESS_OP_TYPE_DESC:
+        mdesc.attr = PBUF_DESC_EXPRESS;
+        break;
+    case XLIO_EXPRESS_OP_TYPE_FILE_ZEROCOPY:
+        mdesc.attr = PBUF_DESC_MDESC;
+        break;
+    default:
+        return -1;
+    };
+    mdesc.mkey = mkey;
+    mdesc.opaque = opaque_op;
+
+    int bytes_written = 0;
+
+    lock_tcp_con();
+
+    for (unsigned i = 0; i < iov_len; ++i) {
+        err = tcp_write_express(&m_pcb, iov[i].iov_base, iov[i].iov_len, &mdesc);
+        if (err != ERR_OK) {
+            return -1;
+        }
+        bytes_written += iov[i].iov_len;
+    }
+
+    if (!(flags & XLIO_EXPRESS_MSG_MORE)) {
+        err = tcp_output(&m_pcb);
+        if (err != ERR_OK) {
+            return -1;
```

Comment:
The same as above, if we create TCP segments with tcp_write() and don't support partial write, we need either do revert to the segments or move socket to error state.

Review what errors tcp_output() can return and if it returns recoverable error - return bytes_written, because we already created all the segments. This may lead to a deadlock in a corner case thoug...

Reply:
> tcp_output() doesn't have to guarantee to send anything to wire